Feature/chargeback

CBothwell opened the pull request:
I was needing access to the chargeback api as well as the fraud service api. I noticed that this was discussed previously in #18 and that a pull request for such a feature would be welcomed. 

I've made several changes to bring in chargeback support. I've broken out the Validation functionality to be more polymorphic and I've added a model for chargeback to conform with the existing behavior.

Additionally, I have fixed a couple bugs. One in xt/author-basic.t to check for the correct exception message on user id failure and a autovivification bug in WebService::MinFraud::Client->_fix_booleans.

Commit:
Chargeback API / #18 - Added support for the chargeback api at https://minfraud.maxmind.com/minfraud/chargeback

In WebService::MinFraud
  Modified POD

In WebService::MinFraud::Client
  Updated Client to support $client->chargeback($request)
  Updated POD

Added WebService::MinFraud::Model::Chargeback
  To keep the interface consistent

In WebService::MinFraud::Validator
In WebService::MinFraud::Validator::Base
In WebService::MinFraud::Validator::Chargeback
In WebService::MinFraud::Validator::Factors
In WebService::MinFraud::Validator::FraudService
In WebService::MinFraud::Validator::Insights
In WebService::MinFraud::Validator::Score
  Updated Validator to support validation of chargebacks
  and keeping the same public interface

In t/WebService/MinFraud/Client.t
In t/WebService/MinFraud/Model/Chargeback.t
In t/WebService/MinFraud/Validator.t
In xt/author-basic.t
  Added tests for changes made to Client.pm, Validator, and
    the addition of the Chargeback.pm model

Commit:
Chargeback API / #18 - Fix documentation in README.md

In README.md
  Updated docuemtantion to match what is in WebService::MinFraud.pm

Commit:
Chargeback API - Addressed code review issues

In README.md and WebService::MinFraud
  Fixed documentation to not introduce try/catch

In WebService::MinFraud::Client
  Fixed bug in _fix_booleans to avoid autovivification

In WebService::MinFraud::Validator
  Changed builders to use a more specific InstanceOf object

In t/WebService/MinFraud/Client.
  Fixed test helper methods to work with chargebacks

Commit:
Chargeback API - Define builder in Base with abstract builder method

In WebService::MinFraud::Validator::Base
  Removed Rx plugin loading in _rx, added abstract builder method

In WebService::MinFraud::Validator::Chargeback
  Added _rx_builder to only add types used by chargeback

In WebService::MinFraud::Validator::FraudService
  Added _rx_builder to add types used by all fraud services

Commit:
Chargeback API - fixed text and tests

In WebService::MinFraud::Validator
  Fixed pod

In t/WebService/MinFraud/Client.t
  Removed unneeded method

Commit:
Chargeback API / #18 - Fixed Failing xt tests

In WebService::MinFraud::Client
  Added missing dependency

In xt/author-basic.t
  Fixed incorrect error message on incorrect user id

Commit:
Chargeback API / #18 - tidied files

In WebService::MinFraud::Client
  Tidied file

Commit:
Chargeback API / #18 Fix spelling issues.

Commit:
Chargeback API / #18 Added FraudServices to .stopwords

horgh commented on README.md:
```diff
@@ -39,10 +39,19 @@ version 1.007000
 This distribution provides an API for the
-[MaxMind minFraud Score, Insights, and Factors web services](https://dev.maxmind.com/minfraud/).
+[MaxMind minFraud Score, Insights, and Factors web services](https://dev.maxmind.com/minfraud/)
+and the [axMind minFraud Chargeback web service](https://dev/maxmind.com/minfraud/chargeback/).
```
There's a minor typo here. (We actually regenerate the README from the main module pod, so it would be fixed when we do a release, but anyway!)

horgh commented on lib/WebService/MinFraud/Client.pm:
```diff
@@ -185,7 +210,10 @@ sub _response_for {
         for my $boolean (@Booleans) {
             my ( $object, $key ) = @{$boolean};
-            next unless exists $content->{$object}{$key};
+            if (   !exists $content->{$object}
```
:+1:

horgh commented on lib/WebService/MinFraud/Model/Chargeback.pm:
```diff
@@ -0,0 +1,36 @@
+      license_key => 'abcdef123456',
+  );
+
+  my $request = { device => { ip_address => '24.24.24.24' } };
```
Should the outer key be removed?

horgh commented on t/WebService/MinFraud/Model/Chargeback.t:
```diff
@@ -0,0 +1,14 @@
+use Test::More 0.88;
+use WebService::MinFraud::Model::Chargeback;
+
+subtest 'Can created a chargeback object' => sub {
```
There's a small typo here.

horgh commented on lib/WebService/MinFraud/Validator/Base.pm:
```diff
@@ -0,0 +1,51 @@
+use Data::Delete 0.05;
+use Carp;
+use Data::Rx;
+use Try::Tiny;
```
It looks like a few includes aren't being used here.

horgh commented on lib/WebService/MinFraud/Validator/Chargeback.pm:
```diff
@@ -0,0 +1,68 @@
+            },
+            maxmind_id => {
+                type   => '//str',
+                length => { 'min' => 1, 'max' => 8 },
```
I believe this must be always 8 characters, so the `min` could reflect that.

horgh commented on lib/WebService/MinFraud/Validator/Chargeback.pm:
```diff
@@ -0,0 +1,68 @@
+            },
+            minfraud_id => {
+                type   => '//str',
+                length => { 'min' => 1, 'max' => 36 },
```
Likewise I believe this should always be exactly 36 characters.

horgh commented on lib/WebService/MinFraud/Client.pm:
```diff
@@ -185,7 +210,10 @@ sub _response_for {
+            }
 
             $content->{$object}{$key}
                 = $content->{$object}{$key}
```
It might be good to update the docs below here as well. In particular, the `REQUEST` and `REQUEST METHODS` sections could reflect the new feature!

horgh reviewed (changes requested):
Great job! Thank you for doing this!

Sorry for the delay in getting back to you.

I have comments about some minor things.

Other than the comments, it would also be awesome if you'd like to add something about this to the changelog file!

Commit:
Chargeback API / #18 - Address code review feedback

In README.md
  Fixed typo

In WebService::MinFraud::Client
  Updated POD documentation to reflect addtion of chargeback api

In WebService::MinFraud::Validator::Chargeback
  Fixed length requirements of maxmind_id
  Fixed lenght requirements of minfraud_id

In WebService::MinFraud::Validator::Base
  Removed unneeded imports

In WebService::MinFraud::Model::Chargeback
  Fixed documentation

In t/WebService/MinFraud/Model/Chargeback.t
  Fixed typo

Commit:
Chargeback API / #18 - Updated Change log

In Changes
    Added information on what was changed

Commit:
Chargeback API / #18 - Address problems with smoke tests

In WebService::MinFraud::Client
  Fixed typo
  Perltidied

horgh commented:
Awesome! Thanks again!
